core/src/main/java/org/apache/iceberg/BaseMetastoreOperations.java

```diff
@@ -0,0 +1,161 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg;
+
+import static org.apache.iceberg.TableProperties.COMMIT_NUM_STATUS_CHECKS;
+import static org.apache.iceberg.TableProperties.COMMIT_NUM_STATUS_CHECKS_DEFAULT;
+import static org.apache.iceberg.TableProperties.COMMIT_STATUS_CHECKS_MAX_WAIT_MS;
+import static org.apache.iceberg.TableProperties.COMMIT_STATUS_CHECKS_MAX_WAIT_MS_DEFAULT;
+import static org.apache.iceberg.TableProperties.COMMIT_STATUS_CHECKS_MIN_WAIT_MS;
+import static org.apache.iceberg.TableProperties.COMMIT_STATUS_CHECKS_MIN_WAIT_MS_DEFAULT;
+import static org.apache.iceberg.TableProperties.COMMIT_STATUS_CHECKS_TOTAL_WAIT_MS;
+import static org.apache.iceberg.TableProperties.COMMIT_STATUS_CHECKS_TOTAL_WAIT_MS_DEFAULT;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+import org.apache.iceberg.exceptions.AlreadyExistsException;
+import org.apache.iceberg.exceptions.CommitFailedException;
+import org.apache.iceberg.util.PropertyUtil;
+import org.apache.iceberg.util.Tasks;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class BaseMetastoreOperations {
+  private static final Logger LOG = LoggerFactory.getLogger(BaseMetastoreOperations.class);
+
+  public void commitHelper(BaseMetadata base, BaseMetadata metadata) {
+    String op = metadata instanceof TableMetadata ? "Table" : "View";
+    String opLowerCase = op.toLowerCase();
+    String opName = opName();
+    // if the metadata is already out of date, reject it
+    if (isStale(base)) {
+      if (base != null) {
+        throw new CommitFailedException("Cannot commit: stale table metadata");
+      } else {
+        // when current is non-null, the table exists. but when base is null, the commit is trying
+        // to create the table
+        throw new AlreadyExistsException("Table already exists: %s", opName());
+      }
+    }
+    // if the metadata is not changed, return early
+    if (base == metadata) {
+      LOG.info("Nothing to commit.");
+      return;
+    }
+
+    long start = System.currentTimeMillis();
+    doCommit(base, metadata);
+    deleteRemovedMetadataFiles(base, metadata);
+    requestRefresh();
+
+    LOG.info(
+        "Successfully committed to {} {} in {} ms",
+        opLowerCase,
+        opName,
+        System.currentTimeMillis() - start);
+  }
+
+  protected abstract String opName();
+
+  protected abstract boolean isStale(BaseMetadata baseMetadata);
+
+  protected abstract void requestRefresh();
+
+  protected abstract void deleteRemovedMetadataFiles(
+      BaseMetadata baseMetadata, BaseMetadata newMetadata);;
+
+  protected void doCommit(BaseMetadata baseMetadata, BaseMetadata newMetadata) {
```

Comment:
We can have a new API with `BaseMetadata`.  Although I am seeing challenge on Backward compatibility. 
Couple of questions I have :
1.  We can't change signature of [doCommit](https://github.com/nk1506/iceberg/pull/10/files#diff-1ae8e9490fe1a4b6da8842c1c313fa57e68a674e186bfeace1c763bee1381faaR124) on tableOperation or ViewOperation for backward compatible. 
2. If we deprecate the original one and create `doCommit(BaseMetadata baseMetadata, BaseMetadata newMetadata)` with Base class it can cause issue on runtime. 
3. Should we rename doCommit on base class? and gradually deprecate ?